kurtosis-devnet/pkg/build/docker.go

```diff
@@ -173,63 +227,110 @@ type templateData struct {
+		state.result = initialImageTag
+		return nil
 	}
 
-	// Prepare template data
+	if err := b.sem.Acquire(ctx, 1); err != nil {
+		log.Printf("Failed to acquire build semaphore for %s: %v", projectName, err)
+		return fmt.Errorf("failed to acquire semaphore: %w", err)
+	}
+	defer b.sem.Release(1)
+
 	data := templateData{
-		ImageTag:    imageTag,
+		ImageTag:    initialImageTag,
 		ProjectName: projectName,
 	}
 
-	// Execute template to get command string
 	var cmdBuf bytes.Buffer
 	if err := b.cmdTemplate.Execute(&cmdBuf, data); err != nil {
-		return "", fmt.Errorf("failed to execute command template: %w", err)
+		log.Printf("Build failed for %s: Failed to execute command template: %v", projectName, err)
+		return fmt.Errorf("failed to execute command template: %w", err)
 	}
+	cmdStr := cmdBuf.String()
+
+	cmd := b.cmdFactory("sh", "-c", cmdStr)
+	var stdoutBuf, stderrBuf bytes.Buffer
+	cmd.SetOutput(&stdoutBuf, &stderrBuf)
+
+	startTime := time.Now()
+	log.Printf("Executing build command for %s: %s", projectName, cmdStr)
+	err := cmd.Run()
+	duration := time.Since(startTime)
 
-	// Create command
-	cmd := b.cmdFactory("sh", "-c", cmdBuf.String())
-	output, err := cmd.CombinedOutput()
 	if err != nil {
-		return "", fmt.Errorf("build command failed: %w\nOutput: %s", err, string(output))
+		log.Printf("Build failed for %s after %s: %v", projectName, duration, err)
+		log.Printf("--- Start Output (stdout) for failed %s ---", projectName)
+		log.Print(stdoutBuf.String())
+		log.Printf("--- End Output (stdout) for failed %s ---", projectName)
+		log.Printf("--- Start Output (stderr) for failed %s ---", projectName)
+		log.Print(stderrBuf.String())
+		log.Printf("--- End Output (stderr) for failed %s ---", projectName)
+		return fmt.Errorf("build command failed: %w", err)
 	}
 
 	dockerClient, err := b.dockerProvider.newClient()
 	if err != nil {
-		return "", fmt.Errorf("failed to create docker client: %w", err)
+		log.Printf("Build command succeeded for %s, but Docker client creation failed: %v", projectName, err)
+		return fmt.Errorf("failed to create docker client: %w", err)
 	}
 
-	ctx := context.Background()
-
-	// Inspect the image to get its ID
-	inspect, _, err := dockerClient.ImageInspectWithRaw(ctx, imageTag)
+	inspect, _, err := dockerClient.ImageInspectWithRaw(ctx, initialImageTag)
 	if err != nil {
-		return "", fmt.Errorf("failed to inspect image: %w", err)
+		log.Printf("Build command succeeded for %s in %s, but failed to inspect image '%s': %v", projectName, duration, initialImageTag, err)
+		log.Printf("Stdout: %s", stdoutBuf.String())
+		log.Printf("Stderr: %s", stderrBuf.String())
+		return fmt.Errorf("build command succeeded but failed to inspect image %s: %w", initialImageTag, err)
 	}
 
-	// Get the short ID (first 12 characters of the SHA256)
-	shortID := strings.TrimPrefix(inspect.ID, "sha256:")[:12]
+	shortID := strings.TrimPrefix(inspect.ID, "sha256:")
+	if len(shortID) > 12 {
+		shortID = shortID[:12]
+	}
```

Comment:
This is used in a few places (including the tests); perhaps we should extract it to a helper function?